### PR TITLE
Amazon Images for Aarch64

### PR DESCRIPTION
**Pull request by thefloweringash**
###### Motivation for this change

AMIs for Amazon's A1 instance type. Following advice from `#nixos-dev`, this adds a hydra job to build the images, and a script that will upload and register the prebuilt images given the store path. The uploader does not require a configured aarch64 builder.

While rewriting the `create-amis.sh` script I only used the `awscli` package. I haven't seen it written anywhere, but it seems like the ec2-* tools do not support current AWS features like session tokens and assume-role based access to s3. I find this version more compatible with my authentication scheme (short lived tokens), but recognize this is a large change from the existing behavior.

Configuration details adapted from https://github.com/NixOS/nixpkgs/issues/52779

Tested booting both "arm64" and "x86_64" images produced from `nix-build nixos/release.nix -A amazon_image`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/maintainers/scripts/ec2/amazon-image.nix
Comment (thefloweringash) [inline nixos/maintainers/scripts/ec2/amazon-image.nix]: `ec2.hvm` is blocked by an assertion, but the implementation still seems to support it. I preserved the existing behavior.

File: nixos/maintainers/scripts/ec2/amazon-image.nix
Comment (thefloweringash) [inline nixos/maintainers/scripts/ec2/amazon-image.nix]: Implemented in the same pattern as `ec2.hvm`, but I don't know the reason for this.

File: nixos/modules/virtualisation/amazon-image.nix
Comment (thefloweringash) [inline nixos/modules/virtualisation/amazon-image.nix]: I don't know if this is a strict requirement, but it reduces the possible configurations to support.

**Comment by memberbetty** [comment]
If you can add some documentation too, which would make it possible to use this in 10 minutes, this is the kind of thing that would make me want to run `master`. 

Great to see arm support coming, regardless.

File: nixos/maintainers/scripts/ec2/create-amis.sh
Comment (thefloweringash) [inline nixos/maintainers/scripts/ec2/create-amis.sh]: `ap-east-1` is new here, and may required additional configuration to enable this region

File: nixos/maintainers/scripts/ec2/create-amis.sh
Comment (thefloweringash) [inline nixos/maintainers/scripts/ec2/create-amis.sh]: I can't tell the home region of the `nixos-amis` bucket, this is probably wrong.

File: nixos/release.nix
Comment (edolstra) [inline nixos/release.nix]: `amazon_image` -> `amazonImage`.

File: nixos/maintainers/scripts/ec2/amazon-image.nix
Comment (edolstra) [inline nixos/maintainers/scripts/ec2/amazon-image.nix]: Is `vpc` a sparse / compressed disk format? For Hydra it's important not to increase the size of the images a lot.

File: nixos/maintainers/scripts/ec2/create-amis.sh
Comment (edolstra) [reply nixos/maintainers/scripts/ec2/create-amis.sh]: I don't think we have to do any region-specific configuration.

File: nixos/maintainers/scripts/ec2/create-amis.sh
Comment (edolstra) [reply nixos/maintainers/scripts/ec2/create-amis.sh]: It's in `eu-west-1` (or `EU` if it needs the old S3 location constraint).

**Review by edolstra — APPROVED** [review]

File: nixos/maintainers/scripts/ec2/amazon-image.nix
Comment (thefloweringash) [reply nixos/maintainers/scripts/ec2/amazon-image.nix]: I believe it's sparse but not compressed, comparisons:
```
└── [ 1193577472]  nixos-amazon-image-19.09pre130979.gfedcba-aarch64-linux.vhd
└── [ 1177550848]  nixos-amazon-image-19.09pre130979.gfedcba-aarch64-linux.qcow2
└── [ 1132745216]  nixos-amazon-image-19.09pre130979.gfedcba-x86_64-linux.vhd
└── [ 1116078080]  nixos-amazon-image-19.09pre130979.gfedcba-x86_64-linux.qcow2
```

File: nixos/maintainers/scripts/ec2/create-amis.sh
Comment (thefloweringash) [reply nixos/maintainers/scripts/ec2/create-amis.sh]: Some regions can be individually disabled and enabled, see the Amazon docs on [Managing AWS Regions][]. I think `ap-east-1` is the only region that this applies to, and is disabled by default.

[Managing AWS Regions]: https://docs.aws.amazon.com/general/latest/gr/rande-manage.html

**Comment by edolstra** [comment]
@thefloweringash Can you resolve the merge conflict? Then I'll merge this. Thanks!

File: nixos/maintainers/scripts/ec2/create-amis.sh
Comment (edolstra) [reply nixos/maintainers/scripts/ec2/create-amis.sh]: Yeah, looks like `ap-east-1` and `me-south-1` are currently disabled.

**Comment by thefloweringash** [comment]
Rebased on current master. I haven't tested it since the rebase, but the rebase itself was straightforward.

**Comment by thefloweringash** [comment]
I've now successfully tested booting both amd64 and aarch64 AMIs.

**Comment by kirelagin** [comment]
I’m trying to use this image and I am importing it manually (since there seem to be no images on the nixos.org website currently), however the `import-image` task fails with

```
"StatusMessage": "ClientError: EFI partition detected. UEFI booting is not supported in EC2.",
```

🤔

**Comment by kirelagin** [comment]
Hm, ok, apparently EFI is specifically for aarch64, and one has to pass `--architecture arm64` to `import-image`, however there is probably something wrong on their end:

```
An error occurred (InvalidParameter) when calling the ImportImage operation: Parameter architecture=arm64 has an invalid format.
```

**Comment by thefloweringash** [comment]
> I am importing it manually

> one has to pass `--architecture arm64` to `import-image`,

Are you using `nixos/maintainers/scripts/ec2/create-amis.sh`? It should [handle this][--architecture] for you. It's not the best interface, but you can modify the top of script to configure your bucket and regions in order to run it locally.

I've created an aarch64 ami with create-amis.sh and successfully booted an a1.medium on it. If this is still failing for you, can you share more details so I can try to reproduce it?

[--architecture]: https://github.com/NixOS/nixpkgs/blob/2901593a1ed33211493e191974a7c21fd144c174/nixos/maintainers/scripts/ec2/create-amis.sh#L234

**Pull request merged**
